upgrade-1.x-to-2.x.sh

```diff
@@ -0,0 +1,437 @@
+	    for binary in $tools_binaries; do
+		log "Installing $binary..."
+		curl -s -L -o $tools_path/$binary $download_uri/$binary
+		chmod 755 $tools_path/$binary
+	    done
+	    ;;
+	*)
+	    log ERROR "Binary type $BINARY_TYPE not supported."
+	    ;;
+    esac
+fi
+
+# Stop docker containers
+log "Stopping all containers..."
+systemctl stop update-resin-supervisor.timer > /dev/null 2>&1
+systemctl stop resin-supervisor > /dev/null 2>&1
+$DOCKER stop $($DOCKER ps -a -q) > /dev/null 2>&1 || true
+log "Removing all containers..."
+$DOCKER rm $($DOCKER ps -a -q) > /dev/null 2>&1 || true
+systemctl stop docker
+
+# Switch connman to use bind mounted state dir
+log "Switching connman to bind mounted state dir..."
+mkdir -p /tmp/connman
+cp -a /var/lib/connman/* /tmp/connman/
+# Versions before 1.20 need this to prevent dropping VPN
+sed -i -e 's/NetworkInterfaceBlacklist=docker,veth,tun,p2p/NetworkInterfaceBlacklist=docker,veth,tun,p2p,resin-vpn/' /etc/connman/main.conf
+mount -o bind /tmp/connman /var/lib/connman
+systemctl restart connman
+
+# Save /resin-data to rootB
+log "Making backup filesystem..."
+mkfs.ext4 -F ${root_dev}p3
+mkdir -p /tmp/backup
+mount ${root_dev}p3 /tmp/backup
+log "Backing up resin-data..."
+(cd /mnt/data; tar -zcf /tmp/backup/resin-data.tar.gz resin-data)
+
+# Unmount p6
+log "Unmounting filesystems..."
+umount /mnt/data
+umount /var/lib/docker
+umount /resin-data
+
+# Save conf contents to /boot
+# XXX
+
+# Unmount p5 if mounted
+if mount | grep /mnt/conf; then
+    umount /mnt/conf
+fi
+
+# Unmount p1
+umount /mnt/boot
+
+log "Creating new partition table stage 1..."
+# Delete partitions 4-6
+parted -s $root_dev rm 6
+parted -s $root_dev rm 5
+parted -s $root_dev rm 4
+
+# Desired layout:
+# Partition	Size		Offset
+# Primary	4MiB		0
+# Boot		40MiB		4MiB
+# RootA		312MiB		44MiB
+# RootB		312MiB		356MiB
+# Extended	Whole disk	668MiB
+# Reserved	4MiB		668MiB
+# State		20MiB		672MiB
+# Reserved	4MiB		692MiB
+# Data		Whole disk	696MiB
+
+# Create extended partition
+parted -s $root_dev mkpart extended 668MiB 100%
+
+# Create state partition
+parted -s $root_dev mkpart logical ext4 672MiB 692MiB
+
+# Create data partition
+# We're going to put btrfs on it initially but using type btrfs
+# corrupts the partition table.
+parted -s $root_dev mkpart logical ext4 696MiB 100%
+
+log "Creating new state and data filesystems..."
+# Create resin-state filesystem
+mkfs.ext4 -F -L resin-state ${root_dev}p5
+
+# Create resin-data filesystem
+mkfs.btrfs -f -L resin-data ${root_dev}p6
+
+# Remount data partition
+mount ${root_dev}p6 /mnt/data
+
+# Copy backup of resin-data
+cp /tmp/backup/resin-data.tar.gz /mnt/data
+
+# Unmount backup
+umount /tmp/backup
+
+log "Creating new partition table stage 2..."
+parted -s $root_dev rm 3
+
+# Resize partition 2 to desired size
+# This uses sectors instead of MiB due to a bug in parted. Assumes 512b sectors
```

Comment:
It's the same bug as the one reported here: https://debbugs.gnu.org/cgi/bugreport.cgi/cgi-bin/bugreport.cgi?bug=23511

Upstream parted doesn't seem to be very actively maintained but I submitted a patch anyway.